dali/operators/reader/loader/filesystem.cc

```diff
@@ -47,168 +47,116 @@ std::string join_path(const std::string &dir, const std::string &path) {
-        file_list.push_back(rel_path);
+#endif
+    std::string fname(entry->d_name);
+    for (auto &filter : filters) {
+      if (fnmatch(filter.c_str(), fname.c_str(), case_sensitive_filter ? 0 : FNM_CASEFOLD) == 0) {
+        files.emplace_back(fname);
+        break;
       }
-      // clean up
-      globfree(&pglob);
     }
   }
   closedir(dir);
+  return true;
 }
 
-
-vector<std::string> traverse_directories(const std::string &file_root, const std::string &filter) {
+vector<std::pair<string, int>> traverse_directories(const std::string &file_root,
+                                                    const std::vector<std::string> &filters,
+                                                    const bool case_sensitive_filter) {
   // open the root
-  DIR *dir = opendir(file_root.c_str());
+  std::vector<std::string> subdirs;
+  bool ret = list_subdirectories(subdirs, file_root);
+  DALI_ENFORCE(ret, "Directory " + file_root + " could not be opened.");
 
-  DALI_ENFORCE(dir != nullptr,
-      "Directory " + file_root + " could not be opened.");
-
-  struct dirent *entry;
+  std::vector<std::pair<std::string, int>> file_label_pairs;
+  std::vector<std::string> files;
+  for (unsigned dir_count = 0; dir_count < subdirs.size(); ++dir_count) {
+    files.clear();
+    auto dirpath = join_path(file_root, subdirs[dir_count]);
+    list_files(files, dirpath, filters, case_sensitive_filter);
+    for (const auto& f : files) {
+      file_label_pairs.push_back({join_path(dirpath, f), dir_count});
+    }
+  }
+  // sort file names as well
+  std::sort(file_label_pairs.begin(), file_label_pairs.end());
+  LOG_LINE  << "read " << file_label_pairs.size() << " files from " << subdirs.size()
+            << "directories\n";
+  return file_label_pairs;
+}
 
-  std::vector<std::string> file_list;
-  std::vector<std::string> entry_name_list;
 
-  // always append the root current directory
-  entry_name_list.push_back(".");
+vector<std::string> traverse_directories(const std::string &file_root, const std::string &filter) {
 
-  // now traverse sub-directories
-  while ((entry = readdir(dir))) {
-    struct stat s;
-    std::string entry_name(entry->d_name);
-    std::string full_path = join_path(file_root, entry_name);
-    int ret = stat(full_path.c_str(), &s);
-    DALI_ENFORCE(ret == 0,
-        "Could not access " + full_path + " during directory traversal.");
-    if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0) continue;
-    if (S_ISDIR(s.st_mode)) {
-      entry_name_list.push_back(entry_name);
+  // open the root
+  std::vector<std::string> subdirs;
+  bool ret = list_subdirectories(subdirs, file_root);
+  subdirs.push_back(".");
+  DALI_ENFORCE(ret, "Directory " + file_root + " could not be opened.");
+
+  std::vector<std::string> files;
+  std::vector<std::string> rel_path_files;
+  for (unsigned dir_count = 0; dir_count < subdirs.size(); ++dir_count) {
+    files.clear();
```

Comment:
This is almost certainly a bug. We iterate over subdirs, populate files and then erase everything except the very last iteration.

Reply:
true, this was copied from my prototype with S3 support and when adapting it I introduced this bug. Good catch